### DISCUSSION TITLE
Localization for User Interface of Identity Provider

### DISCUSSION THREAD

**Discussion by Nilaik** [ory/hydra#2446] [Q&A] [answered]
We are currently implementing the user interface of the identity provider for ory hydra. One of our requirements is that the user interface can be localized to the language which the client is currently using. So for example if the client application is displayed in Japanese language the Login and Consent page should also be displayed in Japanese.

At the moment we are somehow stuck how we can get the localization information from the client to the identity provider, since we have to go over hydra's public endpoint and cannot specify the query of the login or consent path directly. In the reference API of GetLoginRequest und GetConsentRequest we saw that there is a field ui_locales in the oidc_context and were wondering if we can use this field for our purpose. Unfortunately I couldn't find any information on how and where this field is set and used for.

As a last resort we thought of getting the localization information from the request_urls state parameter in which the information could be encoded. However we are not sure if this really is a good solution? Does anyone have dealt with localization of the identity provider's UI before and can give us a hint in the right direction?

**Comment by vinckr** [comment, marked as answer]
I have not dealt with this problem, but maybe I can give you some pointers. 

From the API reference
> UILocales is the End-User'id preferred languages and scripts for the user interface, represented as a
> space-separated list of BCP47 [RFC5646] language tag values, ordered by preference. For instance, the value
> "fr-CA fr en" represents a preference for French as spoken in Canada, then French (without a region designation),
> followed by English (without a region designation). An error SHOULD NOT result if some or all of the requested
> locales are not supported by the OpenID Provider.

It is set in the [OIDC context](https://www.ory.sh/hydra/docs/reference/api/#openidconnectcontext):
```
{
  "acr_values": ["string"],
  "display": "string",
  "id_token_hint_claims": {},
  "login_hint": "string",
  "ui_locales": ["string"]
}
```

 ```
   // Information on the OpenID Connect request - only required to process if your UI should support these values.
    "oidc_context": {"ui_locales": [...], ...},

    // Context is an optional object which can hold arbitrary data. The data will be made available when fetching the
    // consent request under the "context" field. This is useful in scenarios where login and consent endpoints share
    // data.
```

Let me know if that helps, and if we can add something to the documentation to make this clearer.

**Reply by Nilaik** [reply]
Thanks for the quick reply. That sounds great and is exactly what we intend to do. However, it is still not clear to me how the client can set `ui_locales` in the `oidc_context`. We are currently generating an authorize request URL via the Go function `func (c *Config) AuthCodeURL(state string, opts ...AuthCodeOption) string` of the oauht2 go package. This generates a request to the public hydra endpoint, so I assume that it is possible to set the ui_locales as query parameter somehow. Is this somewhere documented which parameters can be set for the request URL to hydra?

**Reply by Nilaik** [reply]
I did some further testing and fount out that you can actually just set the `ui_locales` as query parameter. Using the golang package oauth2 you simply add an AuthCodeOption like `oauth2.SetAuthURLParam("ui_locales", "de-DE")` and hydra makes the ui_locales in the oidc_context available. So the go code for generating the request URL would look something like this:

```
oauth2Conf.AuthCodeURL(
	state,
	oauth2.AccessTypeOffline,
	oauth2.SetAuthURLParam("code_challenge", b64PKCEChallenge),
	oauth2.SetAuthURLParam("code_challenge_method", "S256"),
	oauth2.SetAuthURLParam("ui_locales", "de-DE"),
)
```

**Reply by vinckr** [reply]
Awesome! 
Glad you got it working 💪

**Comment by JapuDCret** [comment]
I struggled to get this working, primarily because I did not understand the answer, but now I do and I want to write this down for any future person:
You can retrieve additional information about the login challenge, whenever you check if it's still valid.
Check the response of https://www.ory.sh/docs/hydra/reference/api#tag/oAuth2/operation/getOAuth2LoginRequest and there you'll see the `oidc_context` with the `ui_locales` attribute

In my case I stored this value in the Spring session store and then wrote a LocaleContextResolver that accessed the session store.
